### PR TITLE
Subscriptions: Toggle subscription on a per-post basis

### LINKED ISSUE #757: Subscriptions: Allow activate/deactivate subscription emails per post
On several occasions this has been requested: a toggle to turn on or off email subscriptions for an individual post, just like you can check/uncheck Publicize connections per post.

Original trac ticket:
* https://plugins.trac.wordpress.org/ticket/1709

### PR DESCRIPTION
**Pull request by dereksmart**
fixes #757

This adds a new checkbox to the post submitbox area that allows the poster to bypass email subscriptions for this post.  It adds some custom post meta and passes it to dotcom where we'll check for it and stop the email if it's set.  

<strong>To Test: </strong>
Publish a new post with and without the checkbox checked.  Should email to subscribers if not checked, and vice-a-versa.  

screen shot 2015-08-19 at 1 22 53 pm

### CODE REVIEW

**Comment by dereksmart** [comment]
@georgestephanis, when you get around to it I would love some eyes on this + my dotcom patch

File: modules/subscriptions.php [hunk 1]
```diff
@@ -114,6 +118,52 @@ function xmlrpc_methods( $methods ) {
...
+	 */
+	function subscription_post_page_metabox() {
+		global $post;
+		$disable_subscribe_value = get_post_meta( $post->ID, '_jetpack_dont_email_post_to_subs', true );
+		// Nonce it
+		wp_nonce_field( 'disable_subscribe', 'disable_subscribe_nonce' );
+		// only show checkbox if post hasn't been published
+		if ( get_post_status( $post->ID ) !== 'publish' ) : ?>
+			<div class="misc-pub-section">
+				<label for="_jetpack_dont_email_post_to_subs">Jetpack Subscriptions: </label><br>
```
Comment (jeherve) [inline modules/subscriptions.php]: We'll need to translate this.

**Comment by jeherve** [comment]
Anything I can do to help test this patch?

**Comment by dereksmart** [comment]
@jeherve yes please!  ping me on slack whenever you're around

**Comment by kraftbj** [comment]
Tested a normal subscription and an excluded one. They were sent out or not sent out respectively as expected. :+1: 

Two stretch goals: 1. A filter to allow a dev to auto-check this exclusion option. 2. Debugging for HEs.

I'll open issues for these. I think it is fine to ship as it is.

**Comment by jeffgolenski** [comment]
Visually, it looks good!

**Comment by gregrickaby** [comment]
@dereksmart @kraftbj @georgestephanis Can this be used to filter out an entire category? 

After WP 4.2, I started using "Press This" as the primary way of organizing my "read it later" bookmarks (instead of a service like Pocket). These posts go into a category named "Press This".

I'm sure my subscribers are wondering why they're getting pinged about recipes and guitar tutorials when my primary blog content is centered around WordPress. 

Thoughts?

**Comment by kraftbj** [comment]
@gregrickaby The technical bits -- the subscription e-mailer on wp.com was updated to ignore any post with the post meta of `'_jetpack_dont_email_post_to_subs' => 1`

I haven't tried it but if you hooked into `wp_insert_post` (or something similar) to automatically add that meta to post with that category, it should work. The sync process that sends subscriptions is intentionally pretty late.

Untested psuedocode:
```add_action( 'wp_insert_post', 'jetpack_subs_be_gone', 10, 4 );
function jetpack_subs_be_done( $post_id, $post, $update ){
$args = array( 'fields
$categories = wp_get_post_categories( $post_id, array( 'fields' => 'names' ) );
if ( in_array( 'Press This', $categories ) {
update_post_meta( $post_id, '_jetpack_dont_email_post_to_subs', 1 );
}
}```

**Comment by gregrickaby** [comment]
@kraftbj Just tested this and it works! Thanks!

```php
	/**
	 * Don't nag my subscribers.
	 *
	 * Attach post meta alerting wordpress.com to ignore this post
	 * and not email my blog subscribers when it's published.
	 * 
	 * @param  integer  $post_id  The post ID.
	 * @param  object   $post     The post object.
	 * @param  bool     $update   Whether this is an existing post being updated or not.
	 */
	public function dont_nag_my_subscribers( $post_id, $post, $update ) {

		// The list of post categories to ignore
		$ignore = array( 'Press This', 'Recipes' );

		// The list of the posts categories associated with this post
		$categories = wp_get_post_categories( $post_id, array( 'fields' => 'names' ) );

		// Compare the two arrays and look for an intersection
		$results = array_intersect( $ignore, $categories );

		// If any of the $ignore items are present, update post meta
		if ( ! empty( $results ) ) {
			update_post_meta( $post_id, '_jetpack_dont_email_post_to_subs', 1 );
		}
	}
	add_action( 'wp_insert_post', 'dont_nag_my_subscribers', 10, 4 );

**Comment by kraftbj** [comment]
Since my last comment, @dereksmart added some more love.
```
/**
		 * Array of categories that will never trigger subscription emails.
		 * Will not send subscription emails from any post from within these categories.
		 *
		 * @since 3.7.0
		 *
		 * @param array $args Array of category slugs or ID's.
		 */
		$excluded_categories = apply_filters( 'jetpack_subscriptions_exclude_these_categories', array() );```

A note that if this filter (or the other filter that only includes particular categories), then the per-post checkbox will not be available. The idea is that either the excluded filter, the included filter, or per-post will be used (at this time, door not closed for future enhancement, but wouldn't be for 3.7).

Can you update to the latest `master` copy of JP and try that? No need for us to debug my suggestion with the built-in option!

**Comment by dereksmart** [comment]
@gregrickaby more info in #2598

**Comment by gregrickaby** [comment]
Thanks @dereksmart and @kraftbj 

See: https://github.com/Automattic/jetpack/pull/2598#issuecomment-134679934

**Pull request merged**